overlord/snapstate/snapstate.go

```diff
@@ -1617,3 +1619,51 @@ func ConfigDefaults(st *state.State, snapName string) (map[string]interface{}, e
 
 	return defaults, nil
 }
+
+func refreshMisc(aStore StoreService) error {
+	e1 := refreshNames(aStore)
+	e2 := refreshSections(aStore)
+	if e1 != nil {
+		return e1
+	}
+	return e2
+}
+
+func refreshNames(aStore StoreService) error {
+	snapCmds, err := aStore.SnapCommands()
+	if err != nil {
+		return err
+	}
+
+	names := make([]string, len(snapCmds))
+	bufsize := 0
+	n := 0
+	for k := range snapCmds {
+		bufsize += len(k)
+		names[n] = k
+		n++
+	}
+	bufsize += n
+	sort.Strings(names)
+	buf := bytes.NewBuffer(make([]byte, 0, bufsize))
+	for _, name := range names {
+		fmt.Fprintln(buf, name)
+	}
+
+	return osutil.AtomicWriteFile(dirs.SnapNamesFile, buf.Bytes(), 0644, 0)
```

Comment:
How about having an atomic buffered *disk* file write API that doesn't need in-memory store of arbitrarily large buffer?

Reply:
I think we do want to move to streaming apis everywhere, but it's a chunk of work